airsonic-main/src/main/java/org/airsonic/player/dao/GenericDaoHelper.java

```diff
@@ -35,4 +43,35 @@ public NamedParameterJdbcTemplate getNamedParameterJdbcTemplate() {
     public DataSource getDataSource() {
         return dataSource;
     }
+
+    @Override
+    public boolean checkpoint() {
+
+        try {
+            Connection conn = DataSourceUtils.getConnection(getJdbcTemplate().getDataSource());
+            DatabaseMetaData meta = conn.getMetaData();
+            String productName = meta.getDatabaseProductName();
+            int productVersion = meta.getDatabaseMajorVersion();
+
+            // HSQLDB (at least version 1) does not handle automatic checkpoints very well by default.
+            // This makes sure the temporary log is actually written to more persistent storage.
+            if (productName.equals("HSQL Database Engine") && (productVersion == 1 || productVersion == 2)) {
+                LOG.info("Performing database checkpoint");
+                getJdbcTemplate().execute("CHECKPOINT DEFRAG");
+                return true;
+            } else {
+                LOG.debug("Database checkpoint not implemented for '" + productName + "'");
+                return false;
+            }
+        }
+
+        // Since this method is a best-effort operation, we don't want to show
+        // a message if the checkpoint failed ; just assume the operation is
+        // unsupported.
+        catch(java.sql.SQLException e) {
+            LOG.debug("An exception occurred during database checkpoint: " + e.toString());
+        }
+
+        return false;
+    }
```

Comment:
I would actually extract this into its own bean as well as this logic is not "Generic" but very specific to hsql. In other words, I would create a new DAO interface specific for checkpointing. The default implementation should be a noop. For Hsql, you would have this logic. Does that make sense?

Reply:
It does, though I'm a bit uncertain on how to do that. If I overwrite the `daoHelper` bean in the legacy application configuration, does that work? Or is there a better way?

Reply:
Overwriting was pretty easy, it should work now. As a bonus I've moved all the legacy-specific logic to the `LegacyHsqlDaoHelper` class so that everything is in the same place!